design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+- [Changing the API](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+- [Deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)
+
+The proposal takes the top-down approach, starting with the high level
+flow that a common feature will traverse through.
+Continuing with actions that need to be taken and timeline suggestions.
+
+Both feature graduation and discontinuation flows are covered.
+Including the implications on users.
+
+Depending on individual topics, follow-up proposals may extend the basic
+points raised in this proposal.
+
+### Terminology
+Feature Gates and feature configuration are often used interchangeable or not differentiated.
+- Feature Gate: A flag that controls the presence or availability of a feature in the cluster.
+- Feature configuration: Cluster or workload level configuration that allows an admin or user
+  (depending on the feature) to control aspects of a feature operation.
+  A common usage is to determine if features are opt-in or opt-out by default.
+
+### Feature Stages
+A feature is expected to pass in the following order through the following stages:
+1. Enhancement proposal.
+2. Implementation.
+3. Release as Alpha (experimental).
+4. Release as Beta (pre-release for evaluation).
+5. Release as General Availability (graduation).
+6. Removal.
+
+Starting from the Alpha release, it can be removed with restrictions that
+depend on the release stage (Alpha, Beta, GA).
+
+[Removal](#removal) of features is widely discussed later
+in this proposal.
+
+#### Enhancement proposal
+As the first step for introducing a new feature, a formal proposal is
+expected to be shared for public review via mailinglist and
+a [design proposal](https://github.com/kubevirt/community/tree/main/design-proposals).
+
+This is the first opportunity to evaluate a new feature.
+The proposal needs to include motivation, goals, implementation details
+and phases. Review the [proposal template](https://github.com/kubevirt/community/blob/main/design-proposals/proposal-template.md)
+for more information.
+
+#### Implementation
+The development work on the feature is expected to include coding,
+testing, integration and documentation.
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+  There are no backward compatability considerations and it can
+  be removed at any time.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
```

Comment:
WDYT about "Alpha field"\"Alpha API" instead of "Feature Gate" ?

Reply:
+1

IOW a Feature Gate can gate different things - an API, a Field, a Behaviour, sometimes an implementation (root vs nonroot). Not only one, but obviously also any combination of multiple of the examples before.

What do we need to close this conversation?

Reply:
I'm trying to understand how the FG works with different API changes. For example:

1. Let's consider changing the way we scale virt-api, or stopping the use of migrations for CPU/memory hot-plug and using another logic instead.

2. Another change could be adding a new field in a status to expose some important information.

Assuming that both 1 and 2 require going through this process, the first case is clear. If I understand correctly, the new behavior will be disabled by default and will only apply once the FG is enabled. But with the second case, what is the FG used for? Would we hide this information if the FG is disabled? Couldn't we just make it clear that this field is alpha and it is not recommended to be used until it is beta/graduated?

Reply:
> But with the second case, what is the FG used for? Would we hide this information if the FG is disabled? Couldn't we just make it clear that this field is alpha and it is not recommended to be used until it is beta/graduated?

Any behavior and fields that are bound to a non-graduate feature are not to be dependent on in production, as they may go away or change, that is the basic definition.

As a feature can destabilize the overall functionality of the project, it is highly recommended to avoid running the logic when the FG is not set, That includes not to report/update the status field.
But this whole thing is really about implementation and guidelines on how a feature should be integrated in a safe manner. I do not think it is about the process this proposal tries to focus on.

It could be a good topic to follow up on.

Reply:
I think that if it makes sense and there is wide agreement, then sometimes we can avoid the feature gate. Sometimes the disablement of new behavior can become complex when it involves multiple components and fields.

Reply:
Why do we have a written language? Because tribes agreed on certain characters in order to exchange information, they also agreed on more semantics around it, i.e. writing from left to right or right to left.

It's conventions that make our lifes easier.
The convention - eventually rule - to have features behind a feature gate makes it easy to understand what features exist.
FG's are the "map" for features in KubeVirt.

Reply:
Okay, I see your point. Although it would feel a bit strange to add feature gates for some APIs and behavior changes, I'm just worried that the restrictions here would be hard requirements. I would prefer to have a note that this rule can be broken under the right circumstances, just like the note with the exceptions to the period range. But I guess if I'm in the minority on that one, we can mark this thread as resolved.